xtext/org.icyphy.linguafranca/src/lib/core/reactor_threaded.c

```diff
@@ -485,6 +485,141 @@ reaction_t* first_ready_reaction() {
  */
 volatile bool __advancing_time = false;
 
+/**
+ * For the specified reaction, if it has produced outputs, insert the
+ * resulting triggered reactions into the reaction queue.
+ * This procedure assumes the mutex lock is not held and grabs
+ * the lock only when it actually inserts something onto the reaction queue.
+ * @param reaction The reaction that has just executed.
+ */
+void schedule_output_reactions(reaction_t* reaction) {
```

Comment:
Considering how long this function has become, I'd suggest to factor out the common code, put it back in `reactor_common.c`, and factor out the threaded/non-threaded-specific code into a separate function to be called from `schedule_output_reactions` in `reactor_common.c`. This way, the comments that are currently written inline can be pulled to the top of this new function so that it's also visible in the generated documentation. This gives is less code duplication, better documentation, all for the price of just one function call.

Reply:
This is done.  It has the side effect of applying the same optimization now in the unthreaded case, which I thought wouldn't make much difference, but it improves the performance of the TimeLimit test by about 15%!

Reply:
Awesome!

Reply:
On AGX, I also see a 15% improvement for the unthreaded version of TimeLimit. However, the threaded version got about 11% worse.

Reply:
I believe the only difference is an extra function call when things have to get queued on the reaction queue... I doubt this would add 11%, so there must be something else going on...

Reply:
I ran the benchmarks again. The 11% reduction in performance appears to have been an anomaly.